vault/resource_token.go

```diff
@@ -0,0 +1,305 @@
+				Description: "The token lease duration.",
+			},
+			"lease_started": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "The token lease started on.",
+			},
+			"accessor": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "The client token accessor.",
+			},
+			"client_token": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "The client token.",
+				Sensitive:   true,
+			},
+		},
+	}
+}
+
+func tokenCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+
+	role := d.Get("role_name").(string)
+
+	iPolicies := d.Get("policies").(*schema.Set).List()
+	policies := make([]string, 0, len(iPolicies))
+	for _, iPolicy := range iPolicies {
+		policies = append(policies, iPolicy.(string))
+	}
+
+	var createRequest = &api.TokenCreateRequest{}
+
+	if len(policies) > 0 {
+		createRequest.Policies = policies
+	}
+
+	if v, ok := d.GetOk("ttl"); ok {
+		createRequest.TTL = v.(string)
+	}
+
+	if v, ok := d.GetOk("explicit_max_ttl"); ok {
+		createRequest.ExplicitMaxTTL = v.(string)
+	}
+
+	if v, ok := d.GetOk("period"); ok {
+		createRequest.Period = v.(string)
+	}
+
+	if v, ok := d.GetOk("no_parent"); ok {
+		createRequest.NoParent = v.(bool)
+	}
+
+	if v, ok := d.GetOk("no_default_policy"); ok {
+		createRequest.NoDefaultPolicy = v.(bool)
+	}
+
+	if v, ok := d.GetOk("display_name"); ok {
+		createRequest.DisplayName = v.(string)
+	}
+
+	if v, ok := d.GetOk("num_uses"); ok {
+		createRequest.NumUses = v.(int)
+	}
+
+	if v, ok := d.GetOk("renewable"); ok {
+		renewable := v.(bool)
+		createRequest.Renewable = &renewable
+	}
+
+	var resp *api.Secret
+	var err error
+
+	if role != "" {
+		log.Printf("[DEBUG] Creating token with role %q", role)
+		resp, err = client.Auth().Token().CreateWithRole(createRequest, role)
+		if err != nil {
+			return fmt.Errorf("error creating token with role %q: %s", role, err)
+		}
+
+		log.Printf("[DEBUG] Created token accessor %q with role %q", resp.Auth.Accessor, role)
+	} else {
+		log.Printf("[DEBUG] Creating token")
+		resp, err = client.Auth().Token().Create(createRequest)
+		if err != nil {
+			return fmt.Errorf("error creating token: %s", err)
+		}
+
+		log.Printf("[DEBUG] Created token accessor %q", resp.Auth.Accessor)
+	}
+
+	d.Set("lease_duration", resp.Auth.LeaseDuration)
+	d.Set("lease_started", time.Now().Format(time.RFC3339))
+	d.Set("client_token", resp.Auth.ClientToken)
+
+	d.SetId(resp.Auth.Accessor)
+
+	return tokenRead(d, meta)
+}
+
+func tokenRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+
+	id := d.Id()
+
+	log.Printf("[DEBUG] Reading token accessor %q", id)
+	resp, err := client.Auth().Token().LookupAccessor(id)
+	if err != nil {
+		log.Printf("[WARN] Token not found, removing from state")
+		d.SetId("")
+		return nil
+	}
+
+	log.Printf("[DEBUG] Read token accessor %q", id)
+
+	if d.Get("renewable").(bool) && tokenCheckLease(d) {
```

Comment:
@bhuisgen bhuisgen can this be configurable? like the duration before expiration, it tries to renew?

Reply:
yop, I meant custom duration to force the renewing

Reply:
Ok I've added the option `renew_duration` to force the renewing before the server-side expiration and the option `renew_increment` to specify a new custom TTL (by default it equals to the token lease). Can you test ?

```
resource "vault_token" "test" {
  role_name = "${vault_token_role.test.name}"
  policies = [ "test" ]
  ttl = "48h" // expire in 2 days (lease_duration = 172800)
  renewable = true // mandatory :)

  // renew options
  renew_duration = 86400 // force renew each day
  renew_increment = 93600 // new TTL of one day + 2h to be safe
}
```

Reply:
@bhuisgen Nice! 👏 Just two things, what about
- `renew_duration` and `renew_increment` forces new resource, imho not wanted behavior - can we get rid of `ForceNew:    true,`? 
- from my pov, it would be better to specify the duration before the expiration to renew the token. If I understand the code correctly, `renew_duration` is now duration after the creation time. I suggest to do something like ACME has with min_days_remaining https://www.terraform.io/docs/providers/acme/r/certificate.html#certificate-renewal

Reply:
I agree with both remarks. I renamed the option to `renew_min_lease` which specify the minimal lease to engage the token renewing. By default or if `renew_min_lease` is set to 0, the token will be only recreated when it has expired.

Reply:
@adam-janis please test.

Reply:
I am just on it :) Will update you very soon

Reply:
@bhuisgen the logic works, but once it tries to renew the token, Terraform crashes.. does following debug log helps?

```
2018-12-17T22:49:49.298+0100 [DEBUG] plugin.terraform-provider-vault_v1.3.1_x4: -----------------------------------------------------
2018-12-17T22:49:49.560+0100 [DEBUG] plugin.terraform-provider-vault_v1.3.1_x4: 2018/12/17 22:49:49 [DEBUG] Vault API Response Details:
2018-12-17T22:49:49.560+0100 [DEBUG] plugin.terraform-provider-vault_v1.3.1_x4: ---[ RESPONSE ]--------------------------------------
2018-12-17T22:49:49.560+0100 [DEBUG] plugin.terraform-provider-vault_v1.3.1_x4: HTTP/2.0 200 OK
2018-12-17T22:49:49.560+0100 [DEBUG] plugin.terraform-provider-vault_v1.3.1_x4: Content-Length: 407
2018-12-17T22:49:49.560+0100 [DEBUG] plugin.terraform-provider-vault_v1.3.1_x4: Cache-Control: no-store
2018-12-17T22:49:49.560+0100 [DEBUG] plugin.terraform-provider-vault_v1.3.1_x4: Content-Type: application/json
2018-12-17T22:49:49.560+0100 [DEBUG] plugin.terraform-provider-vault_v1.3.1_x4: Date: Mon, 17 Dec 2018 21:49:49 GMT
2018-12-17T22:49:49.560+0100 [DEBUG] plugin.terraform-provider-vault_v1.3.1_x4: Server: nginx/1.14.0
2018-12-17T22:49:49.560+0100 [DEBUG] plugin.terraform-provider-vault_v1.3.1_x4: 
2018-12-17T22:49:49.560+0100 [DEBUG] plugin.terraform-provider-vault_v1.3.1_x4: 
2018-12-17T22:49:49.560+0100 [DEBUG] plugin.terraform-provider-vault_v1.3.1_x4: -----------------------------------------------------
2018-12-17T22:49:49.560+0100 [DEBUG] plugin.terraform-provider-vault_v1.3.1_x4: 2018/12/17 22:49:49 [DEBUG] Lease for token accessor "xxxxxx-yyyyyyy-xxxxxx" renewed, new lease duration 0
2018-12-17T22:49:49.563+0100 [DEBUG] plugin.terraform-provider-vault_v1.3.1_x4: 
Error: Error refreshing state: 1 error(s) occurred:

* module.b2b_prod.vault_token.project: 1 error(s) occurred:

* module.b2b_prod.vault_token.project: vault_token.project: unexpected EOF


panic: interface conversion: interface {} is nil, not []interface {}
```

Reply:
@bhuisgen I have submitted review with some changes